File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 5 - Multiple Linear Regression/avinash_polynomial_regression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Mar  9 15:16:20 2018

@author: aerup
"""

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values # make sure X is always matrix
y = dataset.iloc[:, 2].values # Y is always vector

"""# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""


# Fitting Linear Regression  to the dataset

from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Fitting Polynomial Regression to the data set
from sklearn.preprocessing import PolynomialFeatures
pol_reg = PolynomialFeatures(degree = 4)
X_poly = pol_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

# Visulalising  the linear regression results
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg.predict(X), color='green')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Visulalising  the Polynomial regression results
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg_2.predict(pol_reg.fit_transform(X)), color='green')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Predicting a new result with linear regression
lin_reg.predict(6.5)
# Predicting a new result with Polynomial regression
lin_reg_2.predict(pol_reg.fit_transform(6.5))

File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 6 - Polynomial Regression/Polynomial_Regression/avinash_polynomial_regression.R
# Importing data set
data_set = read.csv('Position_Salaries.csv')
data_set = data_set[2:3]

# Fitting Linear Regression to the Training set
lin_regressor = lm(formula = Salary ~ ., 
                   data =  data_set)
summary(lin_regressor)

# Fitting Polynomial Regression to the Training set
data_set$Level2 = data_set$Level^2
data_set$Level3 = data_set$Level^3
data_set$Level4 = data_set$Level^4
poly_regressor = lm(formula = Salary ~ ., 
               data =  data_set)
summary(poly_regressor)

# Visualize Linear Regression Model
library(ggplot2)
ggplot() + 
  geom_point(aes(x = data_set$Level, y = data_set$Salary),
             colour = 'red') +
  geom_line(aes(x = data_set$Level, y = predict(lin_regressor, newdata = data_set)),
            colour = 'green') +
  ggtitle('Truth or Bluff (Linear Regression)') +
  xlab('Level') +
  ylab('Salary')

# Visualize Polynomial Regression Model
ggplot() + 
  geom_point(aes(x = data_set$Level, y = data_set$Salary),
             colour = 'red') +
  geom_line(aes(x = data_set$Level, y = predict(poly_regressor, newdata = data_set)),
            colour = 'green') +
  ggtitle('Truth or Bluff (Linear Regression)') +
  xlab('Level') +
  ylab('Salary')

# Predicting Linear regression result
y_pred = predict(lin_regressor, data.frame(Level = 6.5))

# Predicting Polynomial Regression Model result
y_pred = predict(poly_regressor, data.frame(Level = 6.5,
                                            Level2 = 6.5^2,
                                            Level3 = 6.5^3,
                                            Level4 = 6.5^4))


